### PR TITLE
feat(client): reuse existing storage resources

### LINKED ISSUE #754: Support using existing `Storage` resources to store a new blob.

### PR DESCRIPTION
**Pull request by giac-mysten**
Allows the client to reuse storage resources that are unused in its wallet.

When storing a blob, the client checks if any unused storage resources are present. 
If there are resources with enough storage and valid for enough epochs are found, the one with the _smallest end epoch_ among the ones with the _smallest storage size_ is chosen, in order to get the tightest possible fit to the desired store operation.

In other words, the client:
* filters all storage resources for which `storage_size >= target_size` and `end_epoch >= target_end_epoch`
* sorts by `storage_size` (increasing), breaking ties by looking at `end_epoch` (smallest wins)
* picks the first in the sorting order.

Closes #754

### CODE REVIEW

File: crates/walrus-sui/src/client.rs [hunk 1]
```diff
@@ -430,12 +462,13 @@ impl ContractClient for SuiContractClient {
         let price = self
             .write_price_for_encoded_length(storage.storage_size)
             .await?;
+        let asdf = self.read_client.get_object_ref(storage.id).await?.into();
```
Comment (jpcsmith) [inline crates/walrus-sui/src/client.rs]: Question: What's asdf?

File: crates/walrus-sui/src/client.rs [hunk 2]
```diff
@@ -710,9 +739,50 @@ impl ContractClient for SuiContractClient {
             &[],
         )
         .await?
-        .filter(|blob| include_expired || blob.storage.end_epoch > current_epoch)
+        .filter(|blob| include_expired || blob.storage.end_epoch > epoch_limit)
```
Comment (jpcsmith) [inline crates/walrus-sui/src/client.rs]: Remark: Since epoch_limit is at least zero, is end_epoch always at least 1? (I suppose yes given our special zero-th epoch?) Would it be worth it to define a constant for epoch zero (like `GENESIS_EPOCH`?) in the code to make the fact that it's a special case clear?

File: crates/walrus-sui/src/client.rs [hunk 3]
```diff
@@ -710,9 +739,50 @@ impl ContractClient for SuiContractClient {
...
+            // Sort in descending order, so we can `pop` the best fit as the last element.
+            let size_cmp = b.storage_size.cmp(&a.storage_size);
+            // Break ties by comparing the end epoch, and take the one that is the closest to
+            // `end_epoch`. NOTE: we are already sure that these values are above the minimum.
+            if size_cmp == std::cmp::Ordering::Equal {
+                b.end_epoch.cmp(&a.end_epoch)
+            } else {
+                size_cmp
+            }
+        });
```
Comment (jpcsmith) [inline crates/walrus-sui/src/client.rs]: Nit: This could be done with `reduce` instead of `collect + sort`.

File: crates/walrus-sui/src/client.rs (hunk 1 above)
Comment (giac-mysten) [reply crates/walrus-sui/src/client.rs]: 😬 forgot to rename the variable sorry.

File: crates/walrus-sui/src/client.rs (hunk 2 above)
Comment (giac-mysten) [reply crates/walrus-sui/src/client.rs]: >Since epoch_limit is at least zero, is end_epoch always at least 1?

yes, because it is the exclusive index of the end epoch (so a blob stored for 1 epoch at epoch 2 ends at epoch 3).

>Would it be worth it to define a constant for epoch zero (like GENESIS_EPOCH?) in the code to make the fact that it's a special case clear?

Probably, but is it needed here?

Comment (jpcsmith) [reply crates/walrus-sui/src/client.rs]: No, but I did not understand why this removed all storage resources for epoch zero, as that would be a valid storage resource for a blob being registered in epoch zero. I then understood that because of how we do genesis, that's not possible. But this nuance is not visible in code.

File: crates/walrus-sui/src/client.rs (hunk 3 above)
Comment (giac-mysten) [reply crates/walrus-sui/src/client.rs]: amazing, let me add that!

File: crates/walrus-sui/src/client.rs (hunk 2 above)
Comment (giac-mysten) [reply crates/walrus-sui/src/client.rs]: I see.

I think there is a more subtle point I think. Returning `0` is not really elegant, because `include_expired` is anyway true and then short circuits. I'll think if there is a better way to avoid the confusing `0`.

File: crates/walrus-sui/src/client.rs (hunk 3 above)
Comment (giac-mysten) [reply crates/walrus-sui/src/client.rs]: actually, @amnn just suggested `min_by_key`. even nicer :)

File: crates/walrus-sui/src/client.rs (hunk 2 above)
Comment (giac-mysten) [reply crates/walrus-sui/src/client.rs]: Ah i realize now this is useless. Completely removed.
